### PR TITLE
Add tests for ObjectIdDict

### PR DESCRIPTION
**Pull request by kshyatt**
These are pretty basic, just making sure the basic constructor and `isempty`/`copy`/`similar` work.

### CODE REVIEW

**Comment by kshyatt** [comment]
@JeffBezanson looks like you were the last person to touch `ObjectIdDict` - this look OK?

File: test/dict.jl [hunk 1]
```diff
@@ -374,6 +376,32 @@ let
...
+    @test length(ca) == length(a)
+    @test ca == a
+
+    ca = empty!(ca)
+    @test length(ca) == 0
+end
+
+let d = @inferred ObjectIdDict(i=>i for i=1:3)
+    @test isa(d, ObjectIdDict)
+    @test d == ObjectIdDict(1=>1, 2=>2, 3=>3)
```
Comment (mauro3) [inline test/dict.jl]: Maybe add a `@test eltype(d)==Pair{Any,Any}`?

File: test/dict.jl [hunk 2]
```diff
@@ -374,6 +376,32 @@ let
     Base.show(Base.IOContext(IOBuffer(), :limit => true), a)
 end
 
+let
+    a = ObjectIdDict()
+    a[1] = a
+    a[a] = 2
```
Comment (StefanKarpinski) [inline test/dict.jl]: Shouldn't we be testing something about what's in `a` after this?

File: test/dict.jl [hunk 3]
```diff
@@ -374,6 +376,32 @@ let
...
+
+    sa = similar(a)
+    @test isempty(sa)
+    @test isa(sa, ObjectIdDict)
+
+    ca = copy(a)
+    @test length(ca) == length(a)
+    @test ca == a
```
Comment (StefanKarpinski) [inline test/dict.jl]: Test that they're not the same object? I.e. `ca !=== a`.

File: test/dict.jl [hunk 4]
```diff
@@ -374,6 +376,32 @@ let
...
+
+    ca = empty!(ca)
+    @test length(ca) == 0
```
Comment (StefanKarpinski) [inline test/dict.jl]: Test that `length(a) == 2` still?

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

**Comment by yuyichao** [comment]
Note that if you want to test the difference between `Dict` and `ObjectIdDict`, you can probably check that `ObjectIdDict(1=>2, 1.0=>3)` has two entries where as `Dict(1=>2, 1.0=>3)` has only one.

File: test/dict.jl (hunk 2 above)
Comment (kshyatt) [inline test/dict.jl]: I've addressed all the comments but this one - what's a reasonable looking test for this?

**Review by StefanKarpinski — APPROVED** [review]

File: test/dict.jl (hunk 2 above)
Comment (StefanKarpinski) [inline test/dict.jl]: Maybe something like this:
```jl
@test length(a) == 2
@test 1 in keys(a)
@test a in keys(a)
@test a[1] === a
@test a[a] === 2
```
In other words check that `a` is the object we thing we've constructed: an associative collection with exactly two keys, which are `1` and `a`, mapped to `a` and `2`, respectively.

**Comment by tkelman** [comment]
@yuyichao's suggestion was not added here and would still be a good idea

**Comment by kshyatt** [comment]
@tkelman Doesn't the test I put in at line 395 (struggling to link, I'm on phone, sorry) test that my `ObjectIdDict` has two elements, not one, as @yuyichao suggested? The other statement is one about `Dict`, isn't it?

**Comment by tkelman** [comment]
Yes but the point there was about testing ObjectIdDict-specific behavior, for which you'd want to test with keys that are equal so would be a single key in a Dict, but have different object id so would be separate keys in an ObjectIdDict.

**Comment by kshyatt** [comment]
Ah, right, well then I will open a second PR to incorporate that and @StefanKarpinski's suggestion as well.

**Pull request merged**